### DISCUSSION TITLE
How do I parse an http request from tokio::net::TcpStream into hyper::http:Request?

### DISCUSSION THREAD

**Discussion by dityas** [hyperium/hyper#2810] [Q&A] [answered]
Hey everyone. KInda new to async. I was looking into the documentation and could not find a way to parse a raw HTTP request read from a TCP stream into the HTTP Request struct. Is there a way to do that?

**Comment by seanmonstar** [comment, marked as answer]
If you're looking to use hyper to handle the HTTP for you, then you want to look at [`hyper::server::conn::Http`](https://docs.rs/hyper/latest/hyper/server/conn/index.html).

If you want to implement HTTP yourself, exactly how you decide to parse and encode is up to you. HTTP has a lot of edge-cases, so I don't recommend it. But if you really wanted, you look at using something like the `httparse` crate, and then convert the parsed types into `http::Request`.

**Reply by dityas** [reply]
Thank you very much for your reply. I did take a look at conn::Http. What I am specifically looking for is to send out a custom error page if the request is not parsable. Right now the server does respond with an HTTP error code for malformed requests. I am looking if there is a way to intervene and do a couple of things like logging connection details, modifying the default response, etc. before the error response is sent. Is this something that's doable in the current version?

**Reply by seanmonstar** [reply]
Specifically handling a malformed HTTP request is not a current feature of hyper.
